### PR TITLE
test(py): add pytest-bdd feature tests for pools and replicas

### PR DESCRIPTION
**Pull request by cjones1024**
roughly translated from test/grpc/test_replica.js

resolves CAS-961

### CODE REVIEW

File: test/python/test_pool.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
+
```
Comment (gila) [inline test/python/test_pool.py:1]: IMHO it probably better to put this in a subdir called pool, with its own dockerfile perhaps?

File: test/python/test_pool.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+@pytest.fixture
+def create_image_file():
+    filename = '/tmp/ms0-aio.img'
+    run_cmd(f"rm -f '{filename}'", True)
+    run_cmd(f"truncate -s 64M '{filename}'", True)
+    yield filename
+    run_cmd(f"rm -f '{filename}'", True)
+
+
+@pytest.fixture
```
Comment (gila) [inline test/python/test_pool.py:67]: I see what you are trying to do here, but why not yield in create pool and then have it destroyed? Fixtures are nice but my (short) experience is that they can make things complicated and surprising when not needed.

File: test/python/test_pool.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+def create_pool(mayastor_instance, mayastor_state):
+    mayastor_state['pool'] = mayastor_instance.ms.CreatePool(pb.CreatePoolRequest(name='p0', disks=['malloc:///disk0?size_mb=100']))
+
+
+@when('the user attempts to create a pool specifying a URI representing an aio disk')
+def create_pool_from_aio_disk(mayastor_instance, create_image_file, mayastor_state):
+    try:
+        mayastor_state['pool'] = mayastor_instance.ms.CreatePool(pb.CreatePoolRequest(name='p0', disks=[f"aio://{create_image_file}"]))
+    except Exception as error:
+        mayastor_state['error'] = error
```
Comment (gila) [inline test/python/test_pool.py:89]: I see several of these, why not raise the exception? (i.e test failed).

File: test/python/test_pool.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+@when('the user lists the current pools')
+def list_pools(mayastor_instance, mayastor_state):
+    try:
+        mayastor_state['pool_list'] = mayastor_instance.ms.ListPools(pb.Null(), wait_for_ready=True).pools
+    except Exception as error:
+        mayastor_state['error'] = error
+
+
+@then('the pool creation should fail')
+def pool_creation_should_fail(find_pool, mayastor_state):
```
Comment (gila) [inline test/python/test_pool.py:142]: why not call this when the test should fail rather than storing an error which might be clobbered?

**Review by gila — CHANGES_REQUESTED** [review]

File: test/python/test_pool.py (hunk 1 above)
Comment (cjones1024) [reply test/python/test_pool.py:1]: Yes I think this should happen eventually, but I also think there might be scope for further reorganisation as well.

File: test/python/test_replica.py [hunk 5]
```diff
@@ -0,0 +1,317 @@
+
+import pytest
```
Comment (gila) [inline test/python/test_replica.py:2]: The files do not seem to be properly formatted. I've used `black` on the other files and its also run by the precommit hooks

**Review by gila — CHANGES_REQUESTED** [review]

File: test/python/test_pool.py (hunk 3 above)
Comment (cjones1024) [reply test/python/test_pool.py:89]: Some (although by no means all) tests require the gRPC command to fail so, so raising (or rather not catching) the exception is not always appropriate. The real reason however is an attempt to adhere to the BDD flow: where a 'given' clause sets up the environment, the 'when' clause performs the action under test, and it is the 'then' clause that is responsible for validating the behaviour.

File: test/python/test_pool.py (hunk 2 above)
Comment (cjones1024) [reply test/python/test_pool.py:67]: create_pool is a (given) step function, and is definitely not a fixture. Placing a yield here causes an error (even if nothing happens subsequently. I'm not sure if this is a bug as such or by design.
In any case create_pool is not supposed to destroy the pool as its job is to create the pool in order to test the DestroyPool command.

File: test/python/test_pool.py (hunk 4 above)
Comment (cjones1024) [reply test/python/test_pool.py:142]: As I understand it, this is not really a normal function to be invoked directly, it is much like a fixture and is invoked appropriately via the pytest infrastructure - which ensures that the argument values (again fixtures) are only evaluated once and the results cached appropriately.

File: test/python/test_pool.py [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+        mayastor_state['pool'] = mayastor_instance.ms.CreatePool(pb.CreatePoolRequest(name='p0', disks=[f"aio://{create_image_file}"]))
+    except Exception as error:
+        mayastor_state['error'] = error
+
+
+@when('the user attempts to create a pool specifying a disk with an invalid block size')
+def create_pool_from_disk_with_an_invalid_block_size(mayastor_instance, mayastor_state):
+    try:
+        mayastor_state['pool'] = mayastor_instance.ms.CreatePool(pb.CreatePoolRequest(name='p0', disks=['malloc:///disk0?size_mb=100&blk_size=1024']))
+    except Exception as error:
```
[the remark this answers is not in the document]
Comment (cjones1024) [reply test/python/test_pool.py:96]: There is no straightforward mechanism for passing results from the 'when' step to the 'then' step. Creating a context representing the state of the test allows this communication. It is not Ideal but I have not yet mangaged to come up with a better approach

File: test/python/test_replica.py (hunk 5 above)
Comment (cjones1024) [reply test/python/test_replica.py:1]: reformatted via black as per precommit hook

File: test/python/test_pool.py [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+    "Destroying a pool that does not exist."
+
+
+@scenario("features/pool.feature", "listing pools")
+def test_listing_pools():
+    "Listing pools."
+
+
+@pytest.fixture(scope="module")
+def mayastor_instance(mayastor_mod):
```
Comment (gila) [inline test/python/test_pool.py:45]: Perhaps naming this ms0 is more clear?

File: test/python/test_pool.py [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+
+@pytest.fixture(scope="module")
+def find_pool(mayastor_instance):
+    def lookup(name):
+        for pool in mayastor_instance.ms.ListPools(pb.Null()).pools:
+            if pool.name == name:
+                return pool
+        return None
+
+    yield lookup
```
Comment (gila) [inline test/python/test_pool.py:57]: There is no reason to yield here, only yield what you want something run after the function exits.

File: test/python/test_pool.py [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+    )
+
+
+@when("the user attempts to create a pool specifying a URI representing an aio disk")
+def create_pool_from_aio_disk(mayastor_instance, create_image_file, mayastor_state):
+    mayastor_state["pool"] = mayastor_instance.ms.CreatePool(
+        pb.CreatePoolRequest(name="p0", disks=[f"aio://{create_image_file}"])
+    )
+
+
```
Comment (gila) [inline test/python/test_pool.py:95]: The two functions above both create a pool and one may clobber the other. Why not write a simple function like "get_pool()" and then call the gRPC method?

File: test/python/test_pool.py [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+def create_pool_that_already_exists(mayastor_instance, mayastor_state):
+    mayastor_instance.ms.CreatePool(
+        pb.CreatePoolRequest(name="p0", disks=["malloc:///disk0?size_mb=100"])
+    )
+
+
+@when("the user attempts to destroy a pool that does not exist")
+def destroy_pool_that_does_not_exist(mayastor_instance, mayastor_state):
+    mayastor_instance.ms.DestroyPool(pb.DestroyPoolRequest(name="p0"))
+
```
Comment (gila) [inline test/python/test_pool.py:133]: Why not simply pass pool name?

File: test/python/test_pool.py [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+
+@when("the user attempts to destroy the pool")
+def destroy_pool(mayastor_instance, mayastor_state):
+    mayastor_instance.ms.DestroyPool(
+        pb.DestroyPoolRequest(name=mayastor_state["pool"].name)
+    )
+    del mayastor_state["pool"]
+
```
Comment (gila) [inline test/python/test_pool.py:141]: Same as above 👆

File: test/python/test_pool.py [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+@then("the pool destroy command should succeed")
+@then("the pool should be destroyed")
+def pool_destruction_should_succeed(find_pool):
+    assert find_pool("p0") == None
+
+
+@then("the pool should appear in the output list")
+def pool_should_appear_in_output(mayastor_state):
+    assert mayastor_state["pool"].name in [
+        pool.name for pool in mayastor_state["pool_list"]
```
Comment (gila) [inline test/python/test_pool.py:170]: Again why not call list pool and over that?

**Review by gila — COMMENTED** [review]
@cjones1024 I don't fully get the reasoning behind the cache still. In fact pytest itself already does this where applicable

File: test/python/test_pool.py (hunk 9 above)
Comment (cjones1024) [reply test/python/test_pool.py]: done

**Comment by cjones1024** [comment]
After refactoring, some of the above comments are no longer applicable. Please treat effectively as a new PR

**Review by jonathan-teh — APPROVED** [review]

**Review by gila — APPROVED** [review]

**Comment by cjones1024** [comment]
bors merge

**Pull request merged**